corehq/apps/case_search/utils.py

```diff
@@ -57,24 +60,47 @@ class CaseSearchProfiler:
     timing_context: TimingContext = field(
         default_factory=lambda: TimingContext('Case Search'))
     queries: list = field(default_factory=list)
-    profile_results: list = field(default_factory=list)
+    _query_number: int = 0
 
     def run_query(self, slug, es_query):
+        self._query_number += 1
         if self.debug_mode:
             es_query = es_query.enable_profiling()
 
-        with self.timing_context('run query'):
+        tc = self.timing_context(f'run query #{self._query_number}: {slug}')
+        timer = tc.peek()
+        with tc:
             results = es_query.run()
 
         if self.debug_mode:
-            self.queries.append({'slug': slug, 'query': es_query.raw_query})
-            self.profile_results.append({'slug': slug, 'results': results.raw.get('profile')})
-
+            self.queries.append({
+                'slug': slug,
+                'query_number': self._query_number,
+                'query': es_query.raw_query,
+                'duration': timer.duration,
+                'profile_url': self._get_profile_url(slug, self._query_number, results.raw.get('profile')),
+            })
         return results
 
     def add_query(self, slug, es_query):
+        self._query_number += 1
         if self.debug_mode:
-            self.queries.append({'slug': slug, 'query': es_query.raw_query})
+            self.queries.append({
+                'slug': slug,
+                'query_number': self._query_number,
+                'query': es_query.raw_query,
+                'duration': None,
+                'profile_url': None,
+            })
+
+    @staticmethod
+    def _get_profile_url(slug, query_number, profile_json):
+        timestamp = get_timestamp_for_filename()
+        name = f'es_profile_{query_number}_{slug}_{timestamp}.json'
+        io = BytesIO()
+        io.write(json.dumps(profile_json).encode('utf-8'))
+        io.seek(0)
+        return get_download_url(io, name, content_type='application/json')
```

Comment:
These files will get deleted after 24 hours, based on `timeout` in `get_download_url`, right?

Reply:
that's right